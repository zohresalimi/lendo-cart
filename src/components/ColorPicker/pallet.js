import styled from "styled-components";

export const Pallet = styled.button`
  width: 25px;
  height: 25px;
  border: 1px solid
    ${(props) => (props.color === "white" ? "#ccc" : props.color)};
  border-radius: 50%;
  background-color: ${(props) => props.color};
  &:focus {
    outline: none;
  }
  &:hover {
    cursor: pointer;
    box-shadow: 0 5px 5px rgba(17, 16, 62, 0.15);
  }
`;
