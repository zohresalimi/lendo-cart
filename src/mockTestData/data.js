import React, { Children, useReducer } from "react";
import { BrowserRouter as Router } from "react-router-dom";

import AppContext from "../store/context";
import reducer from "../store/reducer";

export const getTestStore = () => {
  return {
    currentProduct: {
      id: 1,
      selectedFeatures: {
        color: "whith",
        power: 6.5,
      },
    },
    products: {
      byColor: {
        white: {
          1: { power: [6.5, 9.5], remaining: 3 },
          2: { power: [6.5, 9.5], remaining: 3 },
        },
        red: {
          1: { power: [6.5, 9.5], remaining: 7 },
          2: { power: [6.5, 9.5], remaining: 7 },
        },
      },
      byId: {
        1: {
          id: 1,
          name: "Philips hue bulb",
          brand: "Philips",
          price: "500",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
          ],
        },
        2: {
          id: 2,
          name: "TrÃ¥dfria Lampor",
          brand: "IKEA",
          price: "300",
          available: true,
          weight: 0.2,
          options: [
            {
              color: "white",
              power: [6.5, 9.5],
              quantity: 3,
            },
            {
              color: "red",
              power: [6.5, 9.5],
              quantity: 7,
            },
            {
              color: "green",
              power: [6.5],
              quantity: 0,
            },
          ],
        },
      },
    },
    shoppingCart: [
      { id: 1, quantity: 1, selectedFeatures: { color: "white", power: 6.5 } },
    ],
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, getTestStore());

  return (
    <AppContext.Provider
      value={{
        state,
        dispatch: props.mockDispatch || dispatch,
      }}
    >
      <Router>{props.children}</Router>
    </AppContext.Provider>
  );
};
