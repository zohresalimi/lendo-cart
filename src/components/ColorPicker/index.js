import React, { useContext } from "react";
import { Box } from "@material-ui/core";
import DoneIcon from "@material-ui/icons/Done";

import { Pallet } from "./pallet";
import AppContext from "../../store/context";
import { SET_FEATURES } from "../../constant";

import useStyles from "./style";
export default function ColorPicker({ items }) {
  const { state, dispatch } = useContext(AppContext);
  const selectedColor = state.currentProduct.selectedFeatures.color;
  const changeSelectedColor = (color) => {
    dispatch({ type: SET_FEATURES, data: { value: color, name: "color" } });
  };

  const classes = useStyles();
  return (
    <div>
      <div>
        <div>
          <Box component="p">
            Color:
            <Box component="span" mx={0.5}>
              {selectedColor}
            </Box>
          </Box>
        </div>
        <Box display="flex" component="div">
          {items.map((item) => (
            <Box
              key={item}
              component="div"
              mx={0.5}
              position="relative"
              data-testid="select-color"
              onClick={() => changeSelectedColor(item)}
            >
              <DoneIcon
                className={selectedColor === item ? classes.show : classes.hide}
              />
              <Pallet color={item} />
            </Box>
          ))}
        </Box>
      </div>
    </div>
  );
}
