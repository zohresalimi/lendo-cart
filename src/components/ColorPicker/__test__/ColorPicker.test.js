import * as React from "react";

import { act, render, waitFor, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { WithProvider } from "../../../mockTestData/data";

import ColorPicker from "..";

async function renderWrapper(props, mockDispatch) {
  let component;

  act(() => {
    component = render(
      <WithProvider mockDispatch={mockDispatch}>
        <ColorPicker {...props} />
      </WithProvider>
    );
  });

  return component;
}

describe("Color Picker", () => {
  test("should match snapshot", async () => {
    const { container } = await renderWrapper({
      items: ["white", "red"],
    });

    expect(container).toMatchSnapshot();
  });

  test("should change color on button click", async () => {
    const mockDispatch = jest.fn();
    const { getAllByTestId } = await renderWrapper(
      { items: ["white", "red"] },
      mockDispatch
    );
    await waitFor(() => {
      const colorButton = getAllByTestId("select-color")[1];

      fireEvent.click(colorButton);
      expect(mockDispatch).toHaveBeenCalledWith({
        type: "setFeatures",
        data: { value: "red", name: "color" },
      });
    });
  });
});
