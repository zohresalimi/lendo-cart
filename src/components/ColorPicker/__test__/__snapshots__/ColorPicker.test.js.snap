// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color Picker should match snapshot 1`] = `
<div>
  <div>
    <div>
      <div>
        <p
          class="MuiBox-root MuiBox-root-3"
        >
          Color:
          <span
            class="MuiBox-root MuiBox-root-4"
          >
            white
          </span>
        </p>
      </div>
      <div
        class="MuiBox-root MuiBox-root-5"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
          data-testid="select-color"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-show-1"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
            />
          </svg>
          <button
            class="sc-bdfBwQ ejIjHW"
            color="white"
          />
        </div>
        <div
          class="MuiBox-root MuiBox-root-7"
          data-testid="select-color"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-hide-2"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
            />
          </svg>
          <button
            class="sc-bdfBwQ gMWguJ"
            color="red"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
